### PR TITLE
add topk & multitask topk accuracies

### PR DESCRIPTION
**Pull request by XianyuanLiu**
### Description
1. Top-k accuracy for a single task
[`sklearn.metrics.top_k_accuracy_score`](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.accuracy_score.html#sklearn.metrics.accuracy_score) can directly return a top-k accuracy score for each batch. When the number of samples is not divisible by the batch size, the last batch would be shorter than other batches. There would be an error in the final top-k accuracy by averaging each batch's top-k accuracies. In our function, we return the boolean values of each batch and then calculate the final top-k accuracy for every epoch.
2. Top-k accuracy for multi-task

### Status
**Ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [x] In-line docstrings updated.
- [ ] [Source for documentation at `docs`](https://github.com/pykale/pykale/tree/main/docs/source) manually updated for new API.

### CODE REVIEW

File: kale/predict/losses.py [hunk 1]
```diff
@@ -27,6 +27,72 @@ def cross_entropy_logits(linear_output, label, weights=None):
...
 
+def topk_accuracy(output, target, topk=(1,)):
+    """Computes the precision@k for the specified values of k
+
+    Args:
+        output (Tensor): Generated predictions. Shape: [batch_size, class_count].
+        target (Tensor): Ground truth. Shape: [batch_size]
+        topk (tuple(int)): Compute accuracy at top-k for the values of k specified in this parameter.
+    Returns:
+        list(Tensor): A list of tensors of the same length as topk with boolean in.
```
Comment (haipinglu) [inline kale/predict/losses.py:38]: This can be made clearer. 
From the doc, I understand that the output is a list of tensors and the number of tensors is equal to the length of topk.
Please explain more clearly what each tensor is. "with boolean in" is too abstract. What is the size of each tensor? The number of samples? True means in top k and false means not?

File: kale/predict/losses.py [hunk 2]
```diff
@@ -27,6 +27,72 @@ def cross_entropy_logits(linear_output, label, weights=None):
...
+
+    Examples:
+        output = torch.tensor([0.3, 0.2, 0.1], [0.3, 0.2, 0.1])
+        target = torch.tensor((0, 2))
+        top1, top2 = topk_accuracy(output, target, topk=(1, 2))
+    """
+
+    maxk = max(topk)
+    # batch_size = target.size(0)
```
Comment (haipinglu) [inline kale/predict/losses.py:47]: What is the purpose of this commented line? If it is a comment, write in plain text to explain. Otherwise, it is confusing to code readers. Likewise for other commented code below.

File: kale/predict/losses.py [hunk 3]
```diff
@@ -27,6 +27,72 @@ def cross_entropy_logits(linear_output, label, weights=None):
...
+
+    _, pred = output.topk(maxk, 1, True, True)
```
Comment (haipinglu) [inline kale/predict/losses.py:49]: A bit hard to understand here, e.g. why `True`, `True`? Add some explanation may help readers.

File: kale/predict/losses.py [hunk 4]
```diff
@@ -27,6 +27,72 @@ def cross_entropy_logits(linear_output, label, weights=None):
...
+    pred = pred.t()
```
Comment (haipinglu) [inline kale/predict/losses.py:50]: I am lost. What is `.t()`? Convention in Python that I have not picked up?

File: kale/predict/losses.py [hunk 5]
```diff
@@ -27,6 +27,72 @@ def cross_entropy_logits(linear_output, label, weights=None):
...
+    correct = pred.eq(target.view(1, -1).expand_as(pred))
+
+    res = []
```
Comment (haipinglu) [inline kale/predict/losses.py:53]: `res` as result? It can be made more readable.

File: kale/predict/losses.py [hunk 6]
```diff
@@ -27,6 +27,72 @@ def cross_entropy_logits(linear_output, label, weights=None):
...
+    for k in topk:
+        # correct_k = correct[:k].contiguous().view(-1).float().sum(0)
+        # res.append(correct_k.mul_(1.0 / batch_size))
```
Comment (haipinglu) [inline kale/predict/losses.py:56]: Are the above two lines still needed?

File: kale/predict/losses.py [hunk 7]
```diff
@@ -27,6 +27,72 @@ def cross_entropy_logits(linear_output, label, weights=None):
...
+def multitask_topk_accuracy(output, target, topk=(1,)):
+    """Computes the precision@k for the specified values of k for multitask input.
+
+    Args:
+        output (tuple(Tensor)): A tuple of generated predictions. Each tensor is of shape [batch_size, class_count],
+            class_count can vary per task basis, i.e. outputs[i].shape[1] can differ from outputs[j].shape[1].
+        target (tuple(Tensor)): A tuple of ground truth. Each tensor should be of shape [batch_size]
+        topk (tuple(int)): Compute accuracy at top-k for the values of k specified in this parameter.
+    Returns:
+        list(Tensor): A list of tensors of the same length as topk with boolean in.
```
Comment (haipinglu) [inline kale/predict/losses.py:71]: Same as the above

File: kale/predict/losses.py [hunk 8]
```diff
@@ -27,6 +27,72 @@ def cross_entropy_logits(linear_output, label, weights=None):
...
+    return res
+
+
+def multitask_topk_accuracy(output, target, topk=(1,)):
+    """Computes the precision@k for the specified values of k for multitask input.
+
+    Args:
+        output (tuple(Tensor)): A tuple of generated predictions. Each tensor is of shape [batch_size, class_count],
+            class_count can vary per task basis, i.e. outputs[i].shape[1] can differ from outputs[j].shape[1].
+        target (tuple(Tensor)): A tuple of ground truth. Each tensor should be of shape [batch_size]
```
Comment (haipinglu) [inline kale/predict/losses.py:68]: so each tensor is a vector, right?

File: kale/predict/losses.py [hunk 9]
```diff
@@ -27,6 +27,72 @@ def cross_entropy_logits(linear_output, label, weights=None):
...
+        correct_k = torch.ge(correct[:k].float().sum(0), 1)
+        res.append(correct_k)
+    return res
+
+
+def multitask_topk_accuracy(output, target, topk=(1,)):
+    """Computes the precision@k for the specified values of k for multitask input.
+
+    Args:
+        output (tuple(Tensor)): A tuple of generated predictions. Each tensor is of shape [batch_size, class_count],
```
Comment (haipinglu) [inline kale/predict/losses.py:66]: the same shape as in `topk_accuracy` above?

File: kale/predict/losses.py [hunk 10]
```diff
@@ -27,6 +27,72 @@ def cross_entropy_logits(linear_output, label, weights=None):
...
+        all_correct.add_(correct)
+
+    res = []
+    for k in topk:
+        # all_correct_k = all_correct.float().sum(0) * 1.0 / batch_size
+        # res.append(all_correct_k)
+        all_correct_k = torch.ge(all_correct[:k].float().sum(0), task_count)
+        res.append(all_correct_k)
+    return res
+
```
Comment (haipinglu) [inline kale/predict/losses.py:94]: Similar comments to 87 to 93 as the above.

**Review by haipinglu — COMMENTED** [review]
Good job but some tidy-up and clarification are needed before merging

**Comment by XianyuanLiu** [comment]
Besides updating docs and variable names, I made another update—no need to set the device manually for `multitask_topk_accuracy` anymore.

**Review by haipinglu — APPROVED** [review]

**Comment by haipinglu** [comment]
Consider [Torchmetrics](https://torchmetrics.readthedocs.io/en/latest/references/modules.html#id4)
